nemo_curator/modules/semantic_dedup.py

```diff
@@ -0,0 +1,573 @@
+        self.kmeans_with_cos_dist = kmeans_with_cos_dist
+        self.partition_size = partition_size
+        self.sort_clusters = sort_clusters
+        self.logger = self._setup_logger(logger)
+
+        if not os.path.exists(self.clustering_output_dir):
+            expand_outdir_and_mkdir(self.clustering_output_dir)
+        else:
+            self.logger.warning(
+                f"Clustering output directory {self.clustering_output_dir} already exists and will be overwritten"
+            )
+
+    def _setup_logger(self, logger):
+        if isinstance(logger, str):
+            return create_logger(
+                rank=0,
+                name="SemanticClusterLevelDedup",
+                log_file=os.path.join(logger, "SemanticClusterLevelDedup.log"),
+                log_level=logging.INFO,
+                stdout=True,
+            )
+        else:
+            return logger
+
+    def __call__(self, embeddings_dataset: DocumentDataset):
+        embeddings_df = embeddings_dataset.df
+
+        assert "embeddings" in embeddings_df.columns
+        embeddings_df = embeddings_df[[self.id_col, "embeddings"]]
+
+        embeddings_df = embeddings_df.to_backend("pandas").persist()
+        embeddings_df = embeddings_df.repartition(partition_size=self.partition_size)
+        embeddings_df = embeddings_df.to_backend("cudf")
+
+        cupy_darr = embeddings_df.map_partitions(
+            get_embedding_ar, meta=cp.ndarray([1, 1])
+        )
+        cupy_darr.compute_chunk_sizes()
+
+        kmeans = KMeans(n_clusters=self.n_clusters, max_iter=self.max_iter)
+        self.logger.info("KMeans starting fit")
+        kmeans.fit(cupy_darr)
+        self.logger.info("KMeans fit complete")
+
+        self.logger.info(
+            "Computing nearest centroids + distance to centers using kmeans.predict"
+        )
+        nearest_cents = kmeans.predict(cupy_darr)
+        embeddings_df["nearest_cent"] = nearest_cents.astype(np.int32)
+        del nearest_cents
+        meta_df = embeddings_df._meta.copy()
+        meta_df["dist_to_cent"] = cp.zeros(1)
+        embeddings_df = embeddings_df.map_partitions(
+            add_dist_to_cents, centroids=kmeans.cluster_centers_, meta=meta_df
+        )
+        centroids = kmeans.cluster_centers_
+        embeddings_df = embeddings_df.reset_index(drop=True)
+        kmeans_centroids_file = os.path.join(
+            self.clustering_output_dir, "kmeans_centroids.npy"
+        )
+        np.save(kmeans_centroids_file, centroids)
+        self.logger.info("Saving centroids complete")
+        del kmeans, cupy_darr, centroids
+
+        clustering_output_dir = os.path.join(
+            self.clustering_output_dir, "embs_by_nearest_center"
+        )
+        if os.path.exists(clustering_output_dir):
+            self.logger.warning(
+                f"Output directory {clustering_output_dir} already exists and will be overwritten"
+            )
+            shutil.rmtree(clustering_output_dir)
```

Comment:
Not specific to SemDedup, but a general comment: This logic fails when dealing with remote/cloud like data & so it would require additional work to enable this in cloud environments. Not in the scope of this PR but we should track the changes needed to support this with remote storage.

Reply:
This is more of a cleanup guard, We can remove this to a more centralized solution we will end up against. We probably will need a cloud safe version of this in `file_utils` we will import from. 

Can you create an issue regarding "NeMo-Curator Cloud Meta Issue" or something and let's keep adding stuff there.